### DISCUSSION TITLE
Help needed with migration from J2V8

### DISCUSSION THREAD

**Discussion by j0eW** [caoccao/Javet#78] [General]
Hi,

we had our own JS-Parser for our online shop and i was trying to integrate J2V8 i got it running at the end but it was very unstable at server stress tests. I was looking for alternatives and found Javet it looks very promising. I am trying to get it running but can't figure out how to solve my problems.

- is it possible to get method overloading to work with Javet?
- the primitives passed form javascript are correct but my java class is expecting different types is there a way to convert them? For example can the price be Integer or Double and my java class expects Double.

in J2V8 I was registering for each Methode a call Methode from an abstract class that was handling the conversion and the metode overloading based on the amount of parameters and the function name and calling the function on the object at the end.

i hope there is a simple solution to my problemes and I don't have to refactor to much of my codebase.

Best Regards
Jörg Widmann

**Comment by caoccao** [comment]
Thank you for raising these. Yes, Javet supports all the requirements you listed.

Annotation `@V8Function` allows application to specify arbitrary alias name of overloaded function. For instance:
``` java
@V8Function(name = "abc")
public Double getPrice()
```
In JavaScript, `a.abc()` triggers `a.getPrice()` in Java.

Javet is very smart in guessing types. In your case, you just specify you need a `double` and Javet will try to convert `int` to `double` automatically.

Also, Javet allows application to register functions with annotation. It frees developers from manually registering interceptors. Please refer to this [doc](https://github.com/caoccao/Javet/blob/main/docs/reference/v8_function.rst) for detail and let me know if you have any questions.

**Reply by j0eW** [reply]
Hi, thank you for you quick reply. I just tested what you have wrote:

`
@V8Function
public void setPrice(Double price)
`

but i get

com.caoccao.javet.exceptions.JavetExecutionException: Error: Callback signature mismatches: expected parameter type is class java.lang.Double, actual parameter type is class java.lang.Integer

**Reply by caoccao** [reply]
Try this.

``` java
@V8Function public void setPrice(V8Value price)
```

It allows you to bypass the Javet type conversion. You can convert it to whatever type.

**Comment by j0eW** [comment]
**Update

I removed the try block now it is working

---------------------

Hi again, i could fix the problems above. But i got one more, but it should be the last one. How can i pass back a interceptor on a function call and assign it to a variable. I tried it like this but i get 

`com.caoccao.javet.exceptions.JavetExecutionException: Error: Runtime is already registered`

```java
public class V8Product {
  private V8Runtime v8Runtime;
  
  public V8Product(V8Runtime v8Runtime) {
    this.v8runtime = v8Runtime;
  }
  
  @V8Function
  public V8ValueObject addAttribute(String name) {
    V8Attribute attribute = new V8Attribute(v8Runtime, name);
  
//  try (V8ValueObject obj = v8Runtime.createV8ValueObject()) {
    obj.bind(attribute);
    return obj;
//  }
  }
}
```
and the V8Product interceptor is created like this:

```java
V8Product product = new V8Product(v8Runtime);

try (V8ValueObject v8ValueObject = v8Runtime.createV8ValueObject()) {
  v8Runtime.getGlobalObject().set("p", v8ValueObject);
  v8ValueObject.bind(product);
}
```
the V8Attribute class is similar to the V8Product.

The script would then look like this:

```javascript
var attribute1 = p.addAttribute("Test Attribute");
```

thanks in advance
Jörg Widmann

**Reply by caoccao** [reply]
Nice, you get it. Please refer to [Best Practices](https://www.caoccao.com/Javet/reference/tips/best_practices.html).

* Always apply `try-with-resource` to Javet objects regardless of primitive or reference if they are not returned to Javet.
* Always prohibit calling `close()` of Javet objects if they will be returned to Javet.

The rule is simple. In function callback, don't close input arguments, and don't close return value. Javet takes care of the lifecycle.

**Reply by caoccao** [reply]
Also, you don't need to pass the V8Runtime in the constructor. You can do the following and Javet will inject the V8Runtime in `bind()`.

``` java
        // Declare the V8RuntimeSetter for dependency injection.
        @V8RuntimeSetter
        public void setV8Runtime(V8Runtime v8Runtime) {
            this.v8Runtime = v8Runtime;
        }
```

**Comment by caoccao** [comment]
I have prepared a detailed [migration guide](https://www.caoccao.com/Javet/tutorial/migration_guides/migrate_from_j2v8.html).
